### PR TITLE
Fix for Issue #3422

### PR DESCRIPTION
**Pull request by BlaiseD**
Fix for Issue #3422

### CODE REVIEW

**Comment by lbargaoanu** [comment]
Breaking change. And also ad-hoc I'd say. Why would the parent of the last member in the path would be so special? Taking into account the whole path, that would make a lot more sense to me. That would be a breaking change too :) But the whole idea seems kind of subtle and prone to misunderstandings. Not really about the code...

**Comment by lbargaoanu** [comment]
I presume `Include` in EF (kind of similar) works by considering the whole path.

**Comment by BlaiseD** [comment]
Where are you seeing "last member in the path"?  Both of the following set the parent for every item - unless I'm missing something.

```c#
                node.GetMemberExpressions()
                    .Select(e => new MapMemberInfo(e.Member, e.Expression.Type))
```

```c#
                foreach (var memberName in fullMemberName.Split('.'))
                {
                    var currentType = GetCurrentType(property, type);
                    property = currentType.GetFieldOrProperty(memberName);
                    yield return new MapMemberInfo(property, currentType);
                }
```

What would be your approach?  ([requested opinions earlier](https://github.com/AutoMapper/AutoMapper/issues/3422#issuecomment-658244272))

**Comment by BlaiseD** [comment]
A breaking change would break existing tests - right? Or define breaking change.

**Comment by lbargaoanu** [comment]
You interpret the same parameters in a different way. Clearly some code will break. I guess the problem is that the member list is flattened, as opposed to a structural path.

**Comment by BlaiseD** [comment]
Seems to me like it's working as before with exception of the "member belongs to a base class" scenario - which is the fix.

I don't see `MemberInfo.ReflectedType` being used anywhere else in the code. 

The inheritance design is because we collect the member infos before the call to `ExpressionBuilder.GetMapExpression` -  I think changing that interface would have been a bigger break.

It's ok to come up with a new design once you've had time to think about it - this can stay on hold.  We know the underlying problem.

**Review by jbogard — APPROVED** [review]

File: src/AutoMapper/MapMemberInfo.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        public override string Name => _memberInfo.Name;
+
+        public override Type ReflectedType => _parentType;
+
+        public bool Equals(MemberInfo other)
+        {
+            if (ReferenceEquals(null, other)) return false;
+            if (ReferenceEquals(this, other)) return true;
+
+            return this._memberInfo.Equals(other);
```
Comment (lbargaoanu) [inline src/AutoMapper/MapMemberInfo.cs:36]: Always `false`. My co-workers traced a memory leak back to this line. The projection cache keeps growing in 10.1.1. No longer the case now, so it's fixed in 11.0.

Comment (BlaiseD) [reply src/AutoMapper/MapMemberInfo.cs:36]: Yep -  here's from PropertyInfo source:
image

MemberInfo does the same.  Thanks for the heads-up.

**Pull request merged**